## Calculate the standard deviation
To calculate the risk is to calculate the variance of the investment.

In [1]:
import numpy as np
import pandas as pd
from pandas_datareader import data as wb
import matplotlib.pyplot as plt

In [2]:
tickers =['PG','BEI.DE']
sec_data = pd.DataFrame()
for t in tickers:
    sec_data[t] = wb.DataReader(t,data_source='yahoo',start='2007-1-1')['Adj Close']

In [3]:
sec_data.tail()

PG      BEI.DE
Date                              
2022-11-17  140.869995  102.099998
2022-11-18  142.570007  102.349998
2022-11-21  144.380005  103.449997
2022-11-22  145.619995  103.349998
2022-11-23  146.449997  103.900002

In [6]:
#calculate the log return for each day
sec_returns=np.log(sec_data/sec_data.shift(1))

In [5]:
sec_returns

PG    BEI.DE
Date                          
2007-01-03       NaN       NaN
2007-01-04 -0.007621  0.006544
2007-01-05 -0.008624 -0.020772
2007-01-08  0.002202  0.000202
2007-01-09 -0.002517 -0.022858
...              ...       ...
2022-11-17 -0.006157  0.009348
2022-11-18  0.011996  0.002446
2022-11-21  0.012616  0.010690
2022-11-22  0.008552 -0.000967
2022-11-23  0.005684  0.005308

[4003 rows x 2 columns]

In [7]:
#calculate the mean and standard deviation for the year
sec_returns['PG'].mean()*250

0.08016912693332047

In [8]:
sec_returns['PG'].std()*250**0.5

0.1878235437842421

$var*250 = std**2 *250$

$(var*250)**0.5 = std*250**0.5$

In [9]:
sec_returns['BEI.DE'].mean()*250

0.054095033137851946

In [10]:
sec_returns['BEI.DE'].std()*250**0.5

0.21864929800314067

## Quantify the relationship between a diversed investment in a portfolio
correlation coefficient: $\rho xy = \frac{(x-\bar{x})*(y-\bar{y})}{\sigma x\sigma y}$

Covariance = $(x-\bar{x})*(y-\bar{y})$

In [11]:
#calculate the variance
PG_var = sec_returns['PG'].var()
BEI_var = sec_returns['BEI.DE'].var()

In [14]:
PG_var_a = PG_var*250
PG_var_a

0.03527768359967111

In [16]:
BEI_var_a = BEI_var*250
BEI_var

0.0001912300620690649

In [17]:
#calculate the covariance
cov_matrix = sec_returns.cov()
cov_matrix

PG    BEI.DE
PG      0.000141  0.000044
BEI.DE  0.000044  0.000191

In [18]:
cov_matrix_a = sec_returns.cov()*250
cov_matrix_a

PG    BEI.DE
PG      0.035278  0.011021
BEI.DE  0.011021  0.047808

Interpretation: 0.035278 is the variance of PG, and 0.047808 is the variance of BEI.DE. The variance is on the diagonal.

In [19]:
#calculate the correlation coefficient
corr_matrix = sec_returns.corr()
corr_matrix

PG    BEI.DE
PG      1.000000  0.267883
BEI.DE  0.267883  1.000000

Interpretation: the correlation between PG's return and BEI.DE's return is 0.268. Pay attention that the correlation of prices is not the same as the correlation of returns.

## Portfolio variance (2 stocks)
$(w1\sigma 1 + w2\sigma 2)^2$ = $w1^2 \sigma 1^2 +2w1\sigma 1 w2\sigma2 \rho 12 + w2^2 \sigma2^2$

w1+w2 =1

w1 & w 2 are the weights of the 2 stocks

Remember to use the right equation to calculate portfolio variance (count the covariance in)

# Calculate portfolio risk

In [20]:
#set the weights of the 2 stocks
weights = np.array([0.5,0.5])

In [21]:
#calculate portfolio variance (based on linear algebra)
pfolio_var = np.dot(weights.T,np.dot(sec_returns.cov()*250,weights))
pfolio_var

0.026281633522067086

In [22]:
# calculate Portfolio volatility
pfolio_vol = (np.dot(weights.T,np.dot(sec_returns.cov()*250, weights)))**0.5
pfolio_vol

0.16211611123533368

In [25]:
print (str(round(pfolio_vol,5)*100)+ ' %')

16.212 %


## Systematic vs. Idiosyncratic risk
* systematic risk = un-diversifiable risk: depends on the variance of each individual security. Cannot be eliminated.
* idiosyncratic risk = diversifiable risk = company specific risk: driven by company-specific events. Can be eliminated with diversification.
* diversifiable risk = portfolio variance - weighted annual variances

In [26]:
#set the weights
weights = np.array([0.5,0.5])
print(weights[0])
print(weights[1])

0.5
0.5


In [27]:
#diversifiable risk:
PG_var_a

0.03527768359967111

In [28]:
BEI_var_a

0.047807515517266225

In [29]:
dr = pfolio_var - (weights[0]**2*PG_var_a) - (weights[1]**2*BEI_var_a)
dr

0.005510333742832753

In [30]:
#non-diversifiable risk
n_dr_1 = pfolio_var - dr
n_dr_1

0.020771299779234333

In [33]:
# the 2nd method to calculate non-diversifiable risk
n_dr_2 = (weights[0]**2*PG_var_a) +(weights[1]**2*BEI_var_a)
n_dr_2

0.020771299779234333

In [32]:
n_dr_1 == n_dr_2

True